src/ui/components/Overlay/index.js

```diff
@@ -1,80 +1,110 @@
-    children: PropTypes.node,
-    className: PropTypes.string,
-    onEscapeOverlay: PropTypes.func,
-    visibleOnLoad: PropTypes.bool,
-  };
+// TODO: look into this as I don't see this working on the frontend
+// even in other envs.
+export const onClickBackground = (
+  event: SyntheticEvent<HTMLButtonElement>,
+  // eslint-disable-next-line no-use-before-define
+  _this: OverlayBase,
+) => {
+  if (_this.props.onEscapeOverlay) {
+    _this.props.onEscapeOverlay();
+  }
+  _this.hide();
+};
+
+type Props = {|
+  className?: string,
+  children: React.Element<any>,
+  id: string,
+  onEscapeOverlay?: Function,
+  visibleOnLoad?: boolean,
+  _onClickBackground: Function,
+|};
 
+type UIStateType = {|
+  visible: boolean,
+|};
+
+const initialUIState: UIStateType = { visible: false };
+
+type InternalProps = {|
+  ...Props,
+  setUIState: ($Shape<UIStateType>) => void,
+  uiState: UIStateType,
+|};
+
+export class OverlayBase extends React.Component<InternalProps> {
   static defaultProps = {
     visibleOnLoad: false,
+    _onClickBackground: onClickBackground,
   };
 
-  constructor(props) {
-    super(props);
-    this.state = { visible: props.visibleOnLoad };
-  }
+  componentWillReceiveProps(nextProps: InternalProps) {
+    const { uiState: uiStateOld } = this.props;
+    const { visibleOnLoad: visibleOnLoadNew } = nextProps;
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.visibleOnLoad !== undefined) {
-      this.setState({ visible: nextProps.visibleOnLoad });
+    if (
+      (visibleOnLoadNew && !uiStateOld.visible) ||
+      uiStateOld.visible !== nextProps.uiState.visible
+    ) {
+      this.props.setUIState({ visible: visibleOnLoadNew });
     }
   }
 
-  onClickBackground = () => {
-    if (this.props.onEscapeOverlay) {
-      this.props.onEscapeOverlay();
-    }
-    this.hide();
-  };
-
   hide() {
-    this.setState({ visible: false });
+    this.props.setUIState({ visible: false });
   }
 
   show() {
-    this.setState({ visible: true });
+    this.props.setUIState({ visible: true });
   }
 
   toggle() {
-    this.setState({ visible: !this.state.visible });
+    this.props.setUIState({ visible: !this.props.uiState.visible });
   }
 
   render() {
-    const { children, className } = this.props;
+    const { children, className, id, uiState } = this.props;
+
+    invariant(children, 'The children property is required');
+    invariant(id, 'The id property is required');
 
     return (
       <div
         className={makeClassName('Overlay', className, {
-          'Overlay--visible': this.state.visible,
+          'Overlay--visible': uiState.visible,
         })}
-        ref={(ref) => {
-          this.overlayContainer = ref;
-        }}
       >
         <div
-          onClick={this.onClickBackground}
-          ref={(ref) => {
-            this.overlayBackground = ref;
-          }}
           className="Overlay-background"
+          onClick={(e) => this.props._onClickBackground(e, this)}
```

Comment:
You could use [`bind()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Function/bind) instead but it looks like maybe you don't need it as a prop.

Reply:
hmm i don't think bind was working for me but maybe I  don't need it as you suggest. I'll take another look..